### ISSUE TITLE
Bug: no-use-before-define false positives on `function: false` with arrow functions

### ISSUE THREAD

**Issue by ldeninski** [typescript-eslint/typescript-eslint#8513]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.3.3&fileType=.tsx&code=GYCglA3AUMCuB2BjALgSwPbwASjFg3lgL5QwIobbABM4BUWOAzONEdHEmps3YSVESYAzshwAWLAF4sdKQD56jYAFZWUdqSHxROFdNl4FBIkA&eslintrc=N4KABGBEAOCGBOBnApvSAuKABALgT2mUQGN4BLaHAWiIBsyA7HAejiVUgBpwp4BXWkQxhQECJAYB7KnxRUARsgBmk%2BMioATZY2TDIkpUq49xuAkVIVqdRiyky5ilWs3aGuzAG0TYyKniqxmJiosG%2BSnwMxDhkkgyIwkqwtCjcYb7EtLCIKAmYSSnIaelQyAx8ALZ5YAWpPsGQ%2BIRaStW1RfW%2BZADmUmoAKuYASsqoZcRCmDj8HSVQybSSAO4AcrAVyBoAogAe0Ko41dN8yJ0Avj4AujwXZ0A&tsconfig=&tokens=false

### Repro Code

```TypeScript
const f4 = () => {
  f5();
};

const f5 = () => {}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  "rules": {
    "no-use-before-define": "off",
    "@typescript-eslint/no-use-before-define": [
      "error",
      {
        "functions": false,
        "classes": false,
        "enums": false,
        "typedefs": false,
        "ignoreTypeReferences": true,
        "allowNamedExports": true
      }
    ]
  }
};
```

### Expected Result

Call to f5 in f4 should not be marked as error.

### Actual Result

Error: 'f5' was used before it was defined. 10:3 - 10:5

**Comment by auvred** [comment]
I think that `{ "variables": false }` should do the trick!

[Playground](https://typescript-eslint.io/play/#ts=5.3.3&fileType=.tsx&code=GYCglA3AUMCuB2BjALgSwPbwASjFg3lgL5QwIobbABM4BUWOAzONEdHEmps3YSVESYAzshwAWLAF4sdKQD56jYAFZWUdqSHxROFdNl4FBIkA&eslintrc=N4KABGBEAOCGBOBnApvSAuKABALgT2mUQGN4BLaHAWiIBsyA7HAejiVUgBpwp4BXWkQxhQECJAYB7KnxRUARsgBmk%2BMioATZY2TDIkpUq49xuAkVIVqdRiyky5ilWs3aGuzAG0TYyKniqxmJiosG%2BAG4IZLDygojCSrC0KNxhvkp8DMQ4ZJIM8ZiJycg%2BEAC%2BPgC6PBVlQA&tsconfig=&tokens=false)
ESLint docs: https://eslint.org/docs/latest/rules/no-use-before-define#variables

---

The problem is that the function declarations are hoisted, so it's safe to call them even before they are declared. At the same time "arrow functions" are just variables, so they can't be referenced before they are declared.

![We can't call arrow functions before their declaration](https://github.com/typescript-eslint/typescript-eslint/assets/61150013/de102826-2765-4603-8385-a198c066cb7a)

**Comment by ldeninski** [comment]
Indeed. I did not read the docs carefully enough. My apologies.
